### DISCUSSION TITLE
Can one download OpenJFX via unmanaged and have managed jar depend on them?

### DISCUSSION THREAD

**Discussion by hmf** [com-lihaoyi/mill#1105] [Q&A] [answered]
I came across this https://github.com/lihaoyi/mill/issues/767 when trying to use a OpenJX plotting library. The response https://github.com/lihaoyi/mill/issues/767#issuecomment-655926401 shows how I can side-step this temporarily.  I did it as follows:

```scala
  override def unmanagedClasspath = 
      T{
        import coursier._
        //val files = Fetch().addDependencies(dep"org.openjfx:javafx-base:11").run()
        val files = Fetch().addDependencies(dep"eu.hansolo.fx:charts:11.3").run()
        val pathRefs = files.map(f => PathRef(os.Path(f)))
        printf(pathRefs.mkString(","))
        Agg(pathRefs : _*)
      }
```

I also tried simply *downloading* the OpenJFX library and using the charting library via the standard ivy deps. However it still complains about missing dependencies.  Is their any way I can have `ivyDeps` depend on  `unmanagedClasspath`?

Also, I noticed that I get both the Mac and Linux files. This what I get:

```
athRef(/home/hmf/.cache/coursier/v1/https/repo1.maven.org/maven2/eu/hansolo/fx/charts/11.3/charts-11.3.jar,false,-614741131),PathRef(/home/hmf/.cache/coursier/v1/https/repo1.maven.org/maven2/org/openjfx/javafx-base/11/javafx-base-11-mac.jar,false,-710224397),PathRef(/home/hmf/.cache/coursier/v1/https/repo1.maven.org/maven2/org/openjfx/javafx-graphics/11/javafx-graphics-11-mac.jar,false,-779985868),PathRef(/home/hmf/.cache/coursier/v1/https/repo1.maven.org/maven2/org/openjfx/javafx-controls/11/javafx-controls-11-mac.jar,false,569642956),PathRef(/home/hmf/.cache/coursier/v1/https/repo1.maven.org/maven2/org/openjfx/javafx-swing/11/javafx-swing-11-mac.jar,false,135667244),PathRef(/home/hmf/.cache/coursier/v1/https/repo1.maven.org/maven2/org/openjfx/javafx-base/11/javafx-base-11-linux.jar,false,-525694090),PathRef(/home/hmf/.cache/coursier/v1/https/repo1.maven.org/maven2/org/openjfx/javafx-graphics/11/javafx-graphics-11-linux.jar,false,593935270),PathRef(/home/hmf/.cache/coursier/v1/https/repo1.maven.org/maven2/org/openjfx/javafx-base/11/javafx-base-11.jar,false,-703977077),PathRef(/home/hmf/.cache/coursier/v1/https/repo1.maven.org/maven2/org/openjfx/javafx-controls/11/javafx-controls-11-linux.jar,false,-912083503),PathRef(/home/hmf/.cache/coursier/v1/https/repo1.maven.org/maven2/org/openjfx/javafx-graphics/11/javafx-graphics-11.jar,false,-2012466300),PathRef(/home/hmf/.cache/coursier/v1/https/repo1.maven.org/maven2/org/openjfx/javafx-swing/11/javafx-swing-11-linux.jar,false,2080844149)
``` 

Anyway I can avoid this? 

TIA

**Comment by lefou** [comment, marked as answer]
Sorry, but I can't imagine what "it still complains about missing dependencies" really means. Can you more specific and also show the commands you tried and their output?

**Reply by hmf** [reply]
My apologies. I would like to use the [charts Java library](https://github.com/HanSolo/charts). If I use:

```scala
 override def ivyDeps = Agg(
    fansi.withDottyCompat(scalaVersion()),
   betterFiles,
    ...
    //jfxBase,
    charts 
  )
```
the I get the following error:

```
dcase2020.resolvedIvyDeps Failed to load source dependencies
  not found: https://repo1.maven.org/maven2/org/openjfx/javafx-base/11/javafx-base-11-${javafx.platform}.jar
  not found: https://repo1.maven.org/maven2/org/openjfx/javafx-graphics/11/javafx-graphics-11-${javafx.platform}.jar
  not found: https://repo1.maven.org/maven2/org/openjfx/javafx-controls/11/javafx-controls-11-${javafx.platform}.jar
  not found: https://repo1.maven.org/maven2/org/openjfx/javafx-swing/11/javafx-swing-11-${javafx.platform}.jar
```
According to the issue I referred to above, the way to circumvent this is to use the unmaged libraries. So I used:

```scala
override def unmanagedClasspath = 
      T{
        import coursier._
        val files = Fetch().addDependencies(dep"eu.hansolo.fx:charts:11.3").run()
        val pathRefs = files.map(f => PathRef(os.Path(f)))
        printf(pathRefs.mkString(","))
        Agg(pathRefs : _*)
      }
```
and remove the `charts`library from `ivyDeps` and it compiles ok (for example when running tests `./mill -i --watch dcase2020.test data.FilesSpec.*`). What I am asking is, is there anyway I can:

1. add the `charts` and any other libraries that depend on JavaFX to `ivyDeps`,
2. add the JavaFX dependencies via `unmanagedClasspath` such that `ivyDeps` picks up on those libraries. 

In this case I would have:

```scala
override def unmanagedClasspath = 
      T{
        import coursier._
        val files = Fetch().addDependencies(dep"org.openjfx:javafx-base:11").run()
        val pathRefs = files.map(f => PathRef(os.Path(f)))
        printf(pathRefs.mkString(","))
        Agg(pathRefs : _*)
      }
```

for example. I would have to add other jars there too such as graphics, controls and swing.  

It seems like `ivyDeps` does not look at any other libs. Just wanted to know if there is a simple ways that I could let `ivyDeps` know of those jars.

**Reply by lefou** [reply]
Thanks for clarification. The `ivyDeps` target will not look into other places, especially not in `unmanagedClasspath`. What really happens is, that the result of `ivyDeps` is used to resolve and download a classpath. E.g. for `compileClasspath` also the `ivyDeps` targets of other `moduleDeps` will be consulted to build a proper transitive dependency tree.

You could experiment with the `mapDependencies` task (inherited from `CoursierModule`) to "inject" or replace the proper `javafx-platform`. Which is consulted whenever mill resolves managed deps.

**Reply by hmf** [reply]
Thank you for the answer. This will take some digging and experimenting I suppose.  I will cross this bridge when I come to it. Most probably will be back with more questions.

**Reply by lefou** [reply]
More questions are ok. ;-)

You could also try to adapt the solution of PR #775 locally. E.g. create a `MyCoursierModule extends CoursierModule`, apply the change and mix it into your modules.

**Reply by hmf** [reply]
I am lost now - I had assumed that the #767 had not been solved and that #775 only made sure that correct OS specific jars were downloaded. I still get this error on the latest version:

```
dcase2020.resolvedIvyDeps Failed to load source dependencies
  not found: https://repo1.maven.org/maven2/org/openjfx/javafx-base/11/javafx-base-11-${javafx.platform}.jar
```
But I do seem have it downloaded:

```
hmf@xpto:~/.cache/coursier/v1/https/repo1.maven.org/maven2/org/openjfx/javafx-base/11$ ls
javafx-base-11.jar  javafx-base-11-linux.jar  javafx-base-11-mac.jar  javafx-base-11.pom  javafx-base-11-sources.jar
```
Am I missing something? Also note the ${javafx.platform}. Shouldn't that be a value?

**Reply by lefou** [reply]
Hard to tell what's going on here, as I'm not familiar with coursier internals. My assumption is, that coursier properly downloads some referenced artifacts, but can't find the right one afterwards, as we miss the opportunity to replace the platform variable correctly. If you could post your whole `build.sc`, someone with probably more coursier knowledge could
 look into it.

**Comment by hmf** [comment]
@lefou I have set-up a small project here:

  https://github.com/hmf/javaFXMill

Please check the README.md

Tests shows that after version 0.5.3, JavaFX seems to be broken. 

But even then, for the working versions I cannot seem to get a class that is not marked as the main class, to run. JFX complains of missing run-times. 

> If you could post your whole build.sc, someone with probably more coursier knowledge could
look into it.

Are you referring to this thread? Maybe place it elsewhere? Do you mind if I place it in #767 and ask the original poster to at least check the issue related to the non-main class execution failure? 

In addition to this I have tried the workaround with no success. This is a show stopper for me. I am starting to think that this may not be just a problem with Coursier. More concretely I have other projects that use OS specific jars with no problem. This may be a JFX specific issue. Playing around with the module paths did not help either (set as JVM arguments). 

In regards to Coursier, for an attempt at a workaround I reported [this](https://github.com/coursier/coursier/issues/1947). I was thinking, can we have a snapshot with the latest and greatest version of Coursier. Could work from there and experiment to see if we can show that its a Coursier problem?

TIA
